core/chains/evm/gas/models.go

```diff
@@ -337,6 +343,44 @@ func (e *evmFeeEstimator) BumpFee(ctx context.Context, originalFee EvmFee, feeLi
 	return
 }
 
+func (e *evmFeeEstimator) estimateFeeLimit(ctx context.Context, feeLimit uint64, calldata []byte, toAddress *common.Address) (estimatedFeeLimit uint64, err error) {
+	// Use provided fee limit by default if EstimateGasLimit is disabled
+	if !e.geCfg.EstimateGasLimit() {
+		return commonfee.ApplyMultiplier(feeLimit, e.geCfg.LimitMultiplier())
+	}
+	// Create call msg for gas limit estimation
+	// Skip setting Gas to avoid capping the results of the estimation
+	callMsg := ethereum.CallMsg{
+		To:   toAddress,
+		Data: calldata,
+	}
+	estimatedGas, estimateErr := e.ethClient.EstimateGas(ctx, callMsg)
+	if estimateErr != nil {
+		// Do not return error if estimate gas failed, we can still use the provided limit instead since it is an upper limit
+		e.lggr.Errorw("failed to estimate gas limit. falling back to provided gas limit.", "callMsg", callMsg, "providedGasLimit", feeLimit, "error", estimateErr)
+		return feeLimit, nil
+	}
+	e.lggr.Debugw("estimated gas", "callMsg", callMsg, "estimatedGas", estimatedGas, "providedGasLimit", feeLimit)
```

Comment:
Logging the entire `callMsg` will be noisy as it will log the `Data` field as well. I would consider dropping callMsg field altogether as it doesn't provide much value.

Reply:
Good point! Removed it in the latest commit